### PR TITLE
bpo-29782: Use __builtin_clzl for bits_in_digit if available

### PR DESCRIPTION
**Pull request by niklasf**

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow these steps to rectify the issue:

1. Sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/). The "bugs.python.org username" requested by the form is the "Login name" field in ["Your Details" at b.p.o](https://cloud.githubusercontent.com/assets/2680980/23276970/d14a380c-f9d1-11e6-883d-e13b6b211239.png)
2. **Wait at least one US business day** and then check the "Contributor form received entry under "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
3. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

**Review by mdickinson — CHANGES_REQUESTED** [review]
Some comments left on the bpo issue. In brief:
- configuration decisions (deciding whether __builtin_clz is available and does the right thing) should be moved into the configuration machinery.
- the code currently assumes that `digit` is the same as `unsigned long`. That's not always the case.
- once the above issues are addressed, we still need to weigh any speed advantage against the costs in terms of maintainability, portability, readability.

**Comment by mdickinson** [comment]
> the code currently assumes that digit is the same as unsigned long

Ah, no. I see. This should work whenever `digit` is no larger than `unsigned long`, thanks to the implicit type conversion. However, that's probably wasteful in the majority of cases; on many (most?) machines, we should be able to use `clz` rather than `clzl`, and it may be faster. And while it's probably true that on all current platforms `digit` is no larger than `unsigned long`, that's not an assumption we should hardwire in unnecessarily.

**Comment by niklasf** [comment]
* Making an extra version for 16bit ints does not seem to be worth it, so I settled for `unsigned long`.
* Factored out so that `mathmodule.c` can also take advantage of the function.
* I might need help with the configuration process. Checking `__GNUC__` does not seem to be uncommon, but is this the place to do it? Note that `__GNUC__` is also defined by clang.
  I created a new source/header pair which might not be the way to go. Should the header contents go to `pyport.h` as you suggested? Where to put the C fallback?

File: Python/pyintrinsics.c [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#include "Python.h"
+
+#ifndef HAVE_BIT_LENGTH
+static const unsigned char BitLengthTable[32] = {
+    0, 1, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4,
+    5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5
+};
+
+int bit_length(unsigned long d) {
```
Comment (vstinner) [inline Python/pyintrinsics.c]: Please rename the function to _Py_bit_length().

Why the result is signed? Why not unsigned int?

File: Include/pyintrinsics.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#ifndef Py_PYINTRINSICS_H
+#define Py_PYINTRINSICS_H
+
+/* Return the smallest integer k such that n < 2**k, or 0 if n == 0.
+ * Equivalent to floor(lg(x))+1.  Also equivalent to: bitwidth_of_type -
+ * count_leading_zero_bits(x)
+ */
+
+#if defined(__GNUC__)
```
Comment (vstinner) [inline Include/pyintrinsics.h]: Is the builtin available in old GCC versions? Any idea when the function was added?

File: Include/pyintrinsics.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+#define HAVE_BIT_LENGTH
+static inline int bit_length(unsigned long d) {
+    return d ? (8 * sizeof(unsigned long) - __builtin_clzl(d)) : 0;
```
Comment (vstinner) [inline Include/pyintrinsics.h]: would it be possible to avoid the conditional code?

File: Include/pyintrinsics.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+}
+#elif defined(_MSC_VER)
+#define HAVE_BIT_LENGTH
+#include <intrin.h>
```
Comment (vstinner) [inline Include/pyintrinsics.h]: Do we need the following pragma?
#pragma intrinsic(_BitScanReverse)

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Include/pyintrinsics.h (hunk 2 above)
Comment (niklasf) [reply Include/pyintrinsics.h]: [gcc 3.4](https://gcc.gnu.org/onlinedocs/gcc-3.4.6/gcc/Keyword-Index.html#Keyword-Index). I made the guard more precise. It also works for clang.

Btw. while looking how this is usually done in Python I found a typo that would affect gcc 3.x: https://github.com/python/cpython/blob/master/Include/pymacro.h#L56 should say `__GNU*C*_MINOR__`. I'll make a seperate PR. #878.

File: Include/pyintrinsics.h (hunk 3 above)
Comment (niklasf) [reply Include/pyintrinsics.h]: Unfortunately I don't think so. `__builtin_clzl(0)` is undefined or rather has different behavior on various processors.

**Comment by niklasf** [comment]
@haypo Thanks for the review. I updated the PR based on your comments.

**Comment by vstinner** [comment]
http://bugs.python.org/issue29782 has been closed as REJECTED.

**Pull request closed without merging**